src/Connect.js

```diff
@@ -337,59 +336,131 @@ class Connect {
+        break
+      case 'function':
+        this._state = { ...this._state, ...update(this._state) }
+        break
+      case 'string':
+        this._state = { ...this._state, ...JSON.parse(update) }
+        break
+      case 'undefined':
+        break
+      default:
+        throw new Error(`Cannot update state with ${update}`)
+    }
+
+    if (this.publicEncKey && this.pushToken) {
+      this.pushTransport = pushTransport(this.pushToken, this.publicEncKey)
     }
-    return JSON.stringify(connectJSONState)
+
+    // Write to localStorage
+    if (this.storage) {
+      store.set('connectState', JSON.stringify(this._state))
+    } 
   }
 
   /**
-   *  Given string of serialized Connect state, it restores that given state to the Connect
-   *  object which it was called on. You can get the serialized state of a connect object
-   *  by calling the serialize() function.
+   * Load state from local storage and set this instance's state accordingly.
+   * Additionally returns the serialized string for you to manually save the
+   * Connect instance's state somewhere else.
    *
-   *  @param    {String}    str      serialized uPort Connect state
+   * @returns {String} The serialized connect state
    */
-  deserialize(str) {
-    const state = JSON.parse(str)
-    this.address = state.address
-    this.mnid = state.mnid
-    this.did = state.did
-    this.doc = state.doc
-    this.keypair = state.keypair
-    this.pushToken = state.pushToken
-    this.publicEncKey = state.publicEncKey
+  getState() {
+    const serialized = store.get('connectState') 
+    const { address, mnid, did, doc, keypair, pushToken, publicEncKey } = JSON.parse(serialized || '{}')
+    this._state = { address, mnid, did, doc, keypair, pushToken, publicEncKey }
+
+    return JSON.stringify(this._state)
   }
 
   /**
-   *  Gets uPort connect state from browser localStorage and sets on this object
+   * Clear any user-specific state from the browser, (both the Connect instance and localStorage)
+   * effectively logging them out. The keypair (app-instance identity) is preserved
    */
-  getState() {
-    const connectState = store.get('connectState')
-    if (connectState) this.deserialize(connectState)
+  logout() {
+    // Clear explicit state
+    this.setState({
+      did: null, 
+      mnid: null, 
+      address: null, 
+      doc: null, 
+      pushToken: null, 
+      publicEncKey: null
+    })
+
+    // Clear all instance variables with references to current state
+    this.pushTransport = null
   }
 
-    /**
-     *  Writes serialized uPort connect state to browser localStorage at key 'connectState'
-     */
-  setState() {
-    const connectState = this.serialize()
-    store.set('connectState', connectState)
+  /**
+   * Clear the entire state of the connect instance, including the keypair, from memory
+   * and localStorage.  Rebuild this.credentials with a new app-instance identity
+   */
+  reset() {
+    this.setState({
```

Comment:
To avoid duplicated code, could you maybe base it on logout() and then do the keypair separately?

Reply:
Good call, I'll fix that now